src/java.base/share/classes/com/sun/crypto/provider/HkdfKeyDerivation.java

```diff
@@ -0,0 +1,391 @@
+    @Override
+    protected SecretKey engineDeriveKey(String alg,
+                                        KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+
+        if (alg == null || alg.isEmpty()) {
+            throw new InvalidParameterSpecException(
+                "the algorithm for the resultant SecretKey may not be null or"
+                + " empty");
+        }
+
+        return new SecretKeySpec(engineDeriveData(kdfParameterSpec), alg);
+
+    }
+
+    /**
+     * Obtain raw data from a key derivation function.
+     *
+     * @return a derived {@code byte[]}
+     *
+     * @throws InvalidParameterSpecException
+     *     if the information contained within the {@code KDFParameterSpec} is
+     *     invalid or incorrect for the type of key to be derived
+     * @throws UnsupportedOperationException
+     *     if the derived key material is not extractable
+     */
+    @Override
+    protected byte[] engineDeriveData(KDFParameterSpec kdfParameterSpec)
+        throws InvalidParameterSpecException {
+        List<SecretKey> ikms;
+        List<SecretKey> salts;
+        SecretKey inputKeyMaterial;
+        SecretKey salt;
+        SecretKey pseudoRandomKey;
+        byte[] info;
+        int length;
+        // A switch would be nicer, but we may need to backport this before
+        // JDK 17
+        // Also, JEP 305 came out in JDK 14, so we can't declare a variable
+        // in instanceof either
+        if (kdfParameterSpec instanceof HKDFParameterSpec.Extract) {
+            HKDFParameterSpec.Extract anExtract =
+                (HKDFParameterSpec.Extract) kdfParameterSpec;
+            ikms = anExtract.ikms();
+            salts = anExtract.salts();
+            // we should be able to combine these Lists of keys into single
+            // SecretKey Objects,
+            // unless we were passed something bogus or an unexportable P11 key
+            try {
+                inputKeyMaterial = consolidateKeyMaterial(ikms);
+                salt = consolidateKeyMaterial(salts);
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "Issue encountered when combining ikm or salt values into"
+                    + " single keys");
+            }
+            // perform extract
+            try {
+                return hkdfExtract(inputKeyMaterial,
+                                   (salt == null) ? null : salt.getEncoded());
+            } catch (InvalidKeyException ike) {
+                throw new InvalidParameterSpecException(
+                    "an HKDF Extract could not be initialized with the given "
+                    + "key or salt material");
+            } catch (NoSuchAlgorithmException nsae) {
+                // This is bubbling up from the getInstance of the Mac/Hmac.
+                // Since we're defining these values internally, it should be
+                // safe to "eat" this NSAE.
```

Comment:
This could happen if someone filters out the required algorithms. In this case, we usually throw a `ProviderException`. 

There are similar cases below.